File: php/src/Validate/Extensions.php
<?php

declare(strict_types=1);

// SPDX-License-Identifier: MIT

namespace HopTop\Vstar\Validate;

use HopTop\Vstar\Component;
use HopTop\Vstar\Ext\Ext;
use HopTop\Vstar\Generated\Codes;

/**
 * spec/05 §3 and spec/04 -- extension namespace compliance.
 *
 * @internal
 */
final class Extensions
{
    /**
     * One warning per property whose name is neither on the RFC
     * allow-list nor prefixed `X-`.
     *
     * The `X-` classification is delegated to {@see Ext::isExtension()} so
     * there is one answer to "what counts as an extension"; the allow-list
     * stays here because it answers the orthogonal question "is this a
     * known RFC property?", which `ext` has no business knowing.
     *
     * @return list<Diagnostic>
     */
    public static function check(Component $c, string $path): array
    {
        $standard = self::standard();
        $out = [];

        foreach ($c->props as $p) {
            if (isset($standard[strtoupper($p->name)])) {
                continue;
            }

            if (Ext::isExtension($p->name)) {
                continue;
            }

            $out[] = Diagnostic::of(
                Codes::UNKNOWN_PROPERTY,
                'property ' . $p->name . ' is not a known RFC 5545/6350 property and does not use'
                    . ' the X- extension prefix (spec/05 §3, spec/04)',
                $path . '.' . $p->name,
            );
        }

        return $out;
    }

    /**
     * How many properties the generated RFC allow-list carries.
     */
    public static function standardPropertyCount(): int
    {
        return count(Codes::STANDARD_PROPERTIES);
    }

    /**
     * The RFC 5545 §3.7-§3.8 / RFC 6350 §6 allow-list as an uppercased
     * lookup set. The list itself is generated from
     * `spec/registry/standard-properties.json`; this is only the index.
     *
     * @return array<string, true>
     */
    private static function standard(): array
    {
        /** @var array<string, true>|null $index */
        static $index = null;

        if ($index === null) {
            $index = [];

            foreach (Codes::STANDARD_PROPERTIES as $name) {
                $index[strtoupper($name)] = true;
            }
        }

        return $index;
    }
}

File: rs/src/validate/extensions.rs
// SPDX-License-Identifier: MIT

//! spec/05 §3, spec/04 — extension namespace compliance.

use super::{diagnostic, Diagnostic};
use crate::ext;
use crate::generated::codes::{STANDARD_PROPERTIES, UNKNOWN_PROPERTY};
use crate::Component;

/// One warning per property whose name is neither on the RFC
/// allow-list nor prefixed `X-`.
///
/// The `X-` classification is delegated to [`ext::is_extension`] so
/// there is one answer to "what counts as an extension"; the allow-list
/// stays here because it answers the orthogonal question "is this a
/// known RFC 5545/6350 property?", which `ext` has no business knowing.
pub(super) fn check(c: &Component, path: &str) -> Vec<Diagnostic> {
    let mut out = Vec::new();
    for p in &c.props {
        if is_standard(&p.name) || ext::is_extension(&p.name) {
            continue;
        }
        out.push(diagnostic(
            UNKNOWN_PROPERTY,
            format!(
                "property {} is not a known RFC 5545/6350 property and does not use the X- extension prefix (spec/05 §3, spec/04)",
                p.name
            ),
            format!("{path}.{}", p.name),
        ));
    }
    out
}

/// How many properties the generated RFC allow-list carries.
pub(super) fn standard_property_count() -> usize {
    STANDARD_PROPERTIES.len()
}

/// Whether `name` is on the generated RFC 5545/6350 allow-list.
///
/// Comparison is case-insensitive per RFC 5545 §3.1. The scan is linear
/// rather than a binary search on purpose: the generated table's order
/// is the generator's business, and a lookup that silently depends on
/// it would start returning wrong answers the day the registry emits
/// the list in a different order.
fn is_standard(name: &str) -> bool {
    STANDARD_PROPERTIES
        .iter()
        .any(|known| known.eq_ignore_ascii_case(name))
}
